---
title: "Clinic data"
author: "Llu√≠s Revilla Sancho"
date: "`r date()`"
output:
  html_document:
    code_folding: show
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, warning = FALSE, fig.width = 120, fig.height = 50, collapse = TRUE)
library("RGCCA")
library("gliomaData")
library("integration")
library("ggplot2")
library("patchwork")
theme_set(theme_bw())
```

# Introduction

In the clinic  data we used a more complex model but we didn't include interaction within transcriptome.
In this new

```{r introduction}
data(ge_cgh_locIGR, package = "gliomaData")
A <- ge_cgh_locIGR$multiblocks
Loc <- factor(ge_cgh_locIGR$y)
levels(Loc) <- colnames(ge_cgh_locIGR$multiblocks$y)
ncomp <- rep(1, length(A))
# rgcca algorithm using the dual formulation for X1 and X2 
# and the dual formulation for X3
A[[3]] = A[[3]][, -3]
```

# Phenotype

That is the default data used, now  we can use other data relevant to the 
disease:

```{r clinic}
data("clinic", package = "gliomaData")
# Remove any column with an NA or empty fields
keep <- vapply(clinic, function(x){!any(is.na(x) | x == "")}, logical(1L))
clinic <- clinic[, keep]
# Remove any value that is unique in all column or each row
keep <- vapply(clinic, function(x){
  y <- length(unique(x))
   y > 1 & y != length(x)}, logical(1L))
clinic <- clinic[, keep]
colnames(clinic)

# We can use the Sex, the age_an, the localization and grade.OMS
fact2numeric <- function(x){
  as.numeric(levels(x))[x]
}

clinic$sexe <- ifelse(clinic$sexe == "M", 1, 0)
clinic$age_an <- as.numeric(sub(",", ".", as.character(clinic$age_an), fixed = TRUE))
# An outlier of age that we assume that that those are days and not years, 
# so it is divided by 365
clinic$age_an[clinic$age_an > 100] <- clinic$age_an[clinic$age_an > 100]/365.25
Localisation <- sapply(levels(clinic$localisation), function(x){
  as.numeric(x == clinic$localisation)})
```
Now that we have ready our data we add it. Note that we add another change, instead of one numeric value for each localization, we have one column for each and 0 or 1 if present (but we remove one to avoid redundance).
```{r data}
A$y <- as.matrix(clinic[, c("sexe", "age_an", "grade.OMS")])
A$Loc <- as.matrix(Localisation[, 1:2]) # We remove the last column because it is redudant
names(A) <- c("GE", "CGH", "Invariable", "Localisation")
```

# Srhinkage

```{r tau}
(min_shrinkage <- sapply(A, function(x) {
  1 / sqrt(ncol(x))
}))
shrinkage <- c(.071,.2, 1, 1)
```

# Design matrix

Now that we are ready to go, we need to test, all the previous things together
We create a design matrix of dimension 4 with 4 weights per position and the interaction
```{r}
nweights <- 4
weight <- seq(from = 0, to = 1, length.out = nweights)
C_list2 <- weight_design(nweights, 4)
C_list3 <- c(C_list2, 
             lapply(C_list2, function(x){x[1, 1] <- weight[2];x}),
             lapply(C_list2, function(x){x[1, 1] <- weight[3];x}),
             lapply(C_list2, function(x){x[1, 1] <- weight[4];x}))
keep <- check_design(C_list3)
C_list3 <- C_list3[keep]
```

# Evaluation of designs

```{r}
testing <- function(x, type) {
  result.sgcca <- sgcca(A, x, c1 = shrinkage, ncomp = rep(1, length(A)), 
                       scheme = type, verbose = FALSE)
  
  out <- analyze(result.sgcca)
  c(out, "var11" = result.sgcca$C[1, 1])
}
```


```{r}
keep <- vapply(C_list3, correct, logical(1L))
design <- C_list3[keep]
out <- sapply(design, testing, type = "centroid")
out2 <- t(out)
def <- as.data.frame(out2)
def$weights <- as.factor(def$weights)
offset <- is.na(def$AVE_inner)
def <- droplevels(def[!offset, ])
rownames(def) <- seq_len(nrow(def))
saveRDS(def, file = "interaction_clinic.RDS")
```

# Exploring results

We can explore by type of model how accurate is the model by looking the number of weights used.
We see that the most complex model doesn't mean more explanation of the variables.
```{r}
library("dplyr")
def %>% 
  group_by(weights) %>% 
  add_count() %>% 
  dplyr::filter(AVE_inner == max(AVE_inner)) %>% 
  arrange(desc(as.numeric(AVE_inner)), desc(weights)) %>% 
  select(-matches("vs"))
```
We can see that the model with high interaction within gene expression and with localization as well as a low interaction between the CGH and the demographics of the sample
We can look at what happens at the extremes, when the interaction is zero:
```{r}
o <- sapply(def[, 10:15], function(x){
  y <- droplevels(def[x == 0, ])
  y$weights <- as.numeric(as.character(y$weights))
  y[y$AVE_inner == max(y$AVE_inner), , drop = TRUE]
})
o2 <- apply(o, 1, unlist)
as_tibble(unique(o2))
```

We can look at what happens at the extremes, when the interaction is 1:
```{r}
o <- sapply(def[, 10:15], function(x){
  y <- droplevels(def[x == 1, ])
  y$weights <- as.numeric(as.character(y$weights))
  y[y$AVE_inner == max(y$AVE_inner), , drop = TRUE]
})
o2 <- apply(o, 1, unlist)
as_tibble(unique(o2))
```

We can look further to the graphs:
```{r}
def2 <- def[, 1:17] %>% mutate(weight = (var12+var13+var14+var23+var24+var34))
p <- ggplot(def2)
r1 <- p + geom_point(aes(cc1, AVE_inner, col = weights)) + guides(col = FALSE)
r2 <- p + geom_point(aes(cc1, AVE_outer, col = weights)) + guides(col = FALSE)
r3 <- p + geom_point(aes(AVE_inner, AVE_outer, col = weights))
p + geom_point(aes(AVE_inner, AVE_outer, col = weights)) + facet_wrap(~weights)
per12 <- p + 
  geom_point(aes(AVE_inner, var12/weight, col = weights)) + 
  ylim(c(0, 1)) + ylab("% var12")
per13 <- p + 
  geom_point(aes(AVE_inner, var13/weight, col = weights)) + 
  ylim(c(0, 1)) + ylab("% var13")
per14 <- p + 
  geom_point(aes(AVE_inner, var14/weight, col = weights)) + 
  ylim(c(0, 1)) + ylab("% var14")
per23 <- p + 
  geom_point(aes(AVE_inner, var23/weight, col = weights)) + 
  ylim(c(0, 1)) + ylab("% var23")
per24 <- p + 
  geom_point(aes(AVE_inner, var24/weight, col = weights)) + 
  ylim(c(0, 1)) + ylab("% var24")
per34 <- p + 
  geom_point(aes(AVE_inner, var34/weight, col = weights)) + 
  ylim(c(0, 1)) + ylab("% var34")
q1 <- p + geom_point(aes(var12, AVE_outer, col = weights))  + guides(col = FALSE) +
  geom_smooth(aes(var12, AVE_outer, col = weights))
q2 <- p + geom_point(aes(var13, AVE_outer, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var13, AVE_outer, col = weights))
q3 <- p + geom_point(aes(var14, AVE_outer, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var14, AVE_outer, col = weights))
q4 <- p + geom_point(aes(var23, AVE_outer, col = weights))  + guides(col = FALSE) +
  geom_smooth(aes(var23, AVE_outer, col = weights))
q5 <- p + geom_point(aes(var24, AVE_outer, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var24, AVE_outer, col = weights))
q6 <- p + geom_point(aes(var34, AVE_outer, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var34, AVE_outer, col = weights))

s1 <- p + geom_point(aes(var1, AVE_inner, col = weights))  + guides(col = FALSE) +
  geom_smooth(aes(var1, AVE_inner, col = weights))
s2 <- p + geom_point(aes(var2, AVE_inner, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var2, AVE_inner, col = weights))
s3 <- p + geom_point(aes(var3, AVE_inner, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var3, AVE_inner, col = weights))
(r1 + r2 + r3) / (q1 + q2 + q3) / (s1 + s2 + s3)
```

# Testing

```{r}
design.final <- matrix(0, nrow = length(A), ncol = length(A), 
                       dimnames = list(names(A), names(A)))
design.final <- symm(design.final, 
                     def[which.max(def$AVE_inner), grep("var", colnames(def))])
result.sgcca <- sgcca(A, design.final, c1 = shrinkage, ncomp = rep(2, length(A)), 
                       scheme = "centroid", verbose = FALSE)
saveRDS(result.sgcca, "model3_best.RDS")
plot(result.sgcca$Y[[1]][, 1], result.sgcca$Y[[2]][, 1], xlab = "GE", ylab = "CGH", 
     col = clinic$localisation)

```

```{r}
l <- looIndex(size(A))
loo_model <- function(x, model){
  
  RGCCA::sgcca(A = subsetData(A, x),
               C = model, 
               scheme = "centroid", 
               verbose = FALSE, c1 = shrinkage
  )
}

model3 <- diag(4)
model3[1, ] <- 1
model3[2, ] <- 1
model3[, 1] <- 1
model3[, 2] <- 1
diag(model3) <- 0
ncomp <- rep(2, length(A))
out <- sgcca(A, C = model3, c1 = shrinkage, ncomp = ncomp)
saveRDS(out, "model3_sgcca_base.RDS")

result.out <- lapply(l, loo_model, model = model3) 
saveRDS(result.out, "loo-model3.RDS")

result.out <- lapply(l, loo_model, model = design.final) 
saveRDS(result.out, "loo-model3_interaction_best.RDS")
```


# SessionInfo {-}

```{r, echo=FALSE}
devtools::session_info()
```

