---
title: "Weight optimization with interaction and superblock"
author: "Llu√≠s Revilla Sancho"
date: "`r date()`"
output:
  html_document:
    code_folding: show
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, warning = FALSE, fig.width = 120, fig.height = 50, collapse = TRUE)
library("RGCCA")
library("gliomaData")
library("ggplot2")
library("patchwork")
library("integration")
theme_set(theme_bw())
```

# Introduction

In this vignette we explore the relationship between weights (and models) with the outcome of the integration using the glioma data available, with an interaction whitin the gene expression.

```{r introduction}
data(ge_cgh_locIGR, package = "gliomaData")
A <- ge_cgh_locIGR$multiblocks
Loc <-factor(ge_cgh_locIGR$y)
levels(Loc) <-colnames(ge_cgh_locIGR$multiblocks$y)
A[[3]] = A[[3]][, -3]
B = A
B[[3]] =Reduce("cbind", A[1:(length(A)-1)])
# superblock
B[[4]] = A[[3]]
names(B) <- c("GE", "CGH", "Superblock", "Loc")
# Prepare some preliminary design/weights
C <-  matrix(0,ncol = 4, nrow = 4)
dimnames(C) <- list(names(B), names(B))
```

# Original

Here we mimic the original study only using centroid scheme instead

```{r}
nb_block = 4
#length(B)
D = matrix(0, nb_block, nb_block)
D[, nb_block - 1] = 1 ; 
D[nb_block - 1, ] = 1 ;
diag(D) = 0

sgcca.glioma = sgcca(B, D, c1 = c(.071,.2, 1/sqrt(NCOL(B[[3]])) + .2, 1),
    ncomp = c(rep(2, (length(B) - 1)),1),
    scheme = "centroid", scale = TRUE,
    verbose = FALSE)
glioma.superblock <- cbind.data.frame(sgcca.glioma$Y[[3]], Loc, Samples = rownames(sgcca.glioma$Y[[3]]))
glioma.GE <- cbind.data.frame(sgcca.glioma$Y[[1]], Loc, Samples = rownames(sgcca.glioma$Y[[3]]))
ggplot(glioma.superblock) +
  geom_text(aes(comp1, comp2, col = Loc, label = Samples)) +
  ggtitle("Original design", subtitle = "Superblock")
ggplot(glioma.GE) +
  geom_text(aes(comp1, comp2, col = Loc, label = Samples)) +
  ggtitle("Original design", subtitle = "GE")
sgcca.glioma$AVE$AVE_inner
```
We can see that it explains 70% of the variance 

# Design

We create a design matrix of dimension 4 with 4 weights per position and the interaction
```{r}
nweights <- 4
weight <- seq(from = 0, to = 1, length.out = nweights)
C_list2 <- weight_design(nweights, 4)
C_list3 <- c(C_list2, 
             lapply(C_list2, function(x){x[1, 1] <- weight[2];x}),
             lapply(C_list2, function(x){x[1, 1] <- weight[3];x}),
             lapply(C_list2, function(x){x[1, 1] <- weight[4];x}))
keep <- check_design(C_list3)
C_list3 <- C_list3[keep]
```


```{r}
testing <- function(x, type) {
  result.sgcca <- RGCCA::sgcca(B, x, 
                               c1 = c(.071, .2, 1/sqrt(NCOL(B[[3]])) + .2, 1), 
                               ncomp = c(1, 1, 1), 
                       scheme = type, verbose = FALSE)
  
  ind <- apply(which(upper.tri(result.sgcca$C), arr.ind = TRUE), 1, 
               paste0, collapse = "")
  
  # Correlation between Y
  Y <- simplify2array(result.sgcca$Y, higher = FALSE)
  cY <- cor(Y)
  d <- cY * x
  vs <- cY[upper.tri(cY)]
  names(vs) <- paste0("vs", ind)
  # Canonical correlation
  cc <- sum(abs(d[upper.tri(d)]))
  # Values of the design matrix
  vars <- result.sgcca$C[upper.tri(result.sgcca$C)]
  names(vars) <- paste0("var", ind)
  vars <- c(vars, "var11" = result.sgcca$C[1, 1])
  weight <- sum(vars != 0)
  names(weight) <- "weights"
  
  # Output
  c(vs, unlist(result.sgcca$AVE[c("AVE_inner", "AVE_outer")]), cc1 = cc, 
    vars, weight)
}
```

# Testing

Using the centroid scheme

```{r centroid}
out <- sapply(C_list3, testing, type = "centroid", USE.NAMES = FALSE)
out2 <- t(out)
def <- as.data.frame(out2)
def$weights <- as.factor(def$weights)
offset <- is.na(def$AVE_inner)
def <- droplevels(def[!offset, ])
saveRDS(def, file = "superblock_interactions.RDS")
```

And we can see the stability for each type of model
```{r}
library("dplyr")
def %>% 
  group_by(weights) %>% add_count() %>% 
  dplyr::filter(AVE_inner == max(AVE_inner)) %>% 
  arrange(desc(AVE_inner), as.numeric(weights)) %>% 
  select(-matches("vs"))
```

We can plot them

```{r plot_centroid}
p <- ggplot(def)
r1 <- p + geom_point(aes(cc1, AVE_inner, col = weights)) + guides(col = FALSE)
r2 <- p + geom_point(aes(cc1, AVE_outer, col = weights)) + guides(col = FALSE)
r3 <- p + geom_count(aes(AVE_inner, AVE_outer, col = weights))
q1 <- p + geom_point(aes(var1, AVE_outer, col = weights))  + guides(col = FALSE) +
  geom_smooth(aes(var2, AVE_outer, col = weights))
q2 <- p + geom_point(aes(var2, AVE_outer, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var2, AVE_outer, col = weights))
q3 <- p + geom_point(aes(var3, AVE_outer, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var3, AVE_outer, col = weights))
s1 <- p + geom_point(aes(var1, AVE_inner, col = weights))  + guides(col = FALSE) +
  geom_smooth(aes(var1, AVE_inner, col = weights))
s2 <- p + geom_point(aes(var2, AVE_inner, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var2, AVE_inner, col = weights))
s3 <- p + geom_point(aes(var3, AVE_inner, col = weights)) + ylab("") + guides(col = FALSE) +
  geom_smooth(aes(var3, AVE_inner, col = weights))
(r1 + r2 + r3) / (q1 + q2 + q3) / (s1 + s2 + s3)
```

We can see the influence of each weight on the inner AVE via modeling it.
```{r}
library("broom")
lmS0 <- lm(AVE_inner~0+var12+var13+var23+var14+var23+var34+var11, data = def)
lmS <- lm(AVE_inner~0+var12*var13*var23*var14*var23*var34*var11, data = def)
anova(lmS0, lmS)
glance(lmS)
tidy(lmS) %>% arrange(desc(abs(estimate)))
lmS1 <- lm(AVE_inner~0+var12+var13+var23+var14+var23+var34+var11, data = def[def$var11 == 0, ])
lmS1b <- lm(AVE_inner~0+var12*var13*var23*var14*var23*var34*var11, data = def[def$var11 == 0, ])
anova(lmS1, lmS1b)
glance(lmS1)
tidy(lmS1) %>% arrange(desc(abs(estimate)))
```

# Final model

We select the best model

```{r}
nb_block = 4
#length(B)
D = matrix(0, nb_block, nb_block)
D[2, 3] <- D[3, 2] <- D[1, 4] <- D[4, 1] <- 1/3
D[1, 1] <- 1
dimnames(D) <- list(names(B), names(B))

sgcca.final = sgcca(B, D, c1 = c(.071,.2, 1/sqrt(NCOL(B[[3]])) + .2, 1),
    ncomp = c(rep(2, (length(B) - 1)),1),
    scheme = "centroid", scale = TRUE,
    verbose = FALSE)
levels(Loc) <- c("hemisphere", "DIPG", "midline")
final.superblock <- cbind.data.frame(sgcca.final$Y[[3]], Loc, Samples = rownames(sgcca.final$Y[[3]]))
final.GE <- cbind.data.frame(sgcca.final$Y[[1]], Loc, Samples = rownames(sgcca.final$Y[[3]]))
ggplot(final.superblock) +
  geom_text(aes(comp1, comp2, col = Loc, label = Samples)) +
  ggtitle("Final design", subtitle = "Superblock") +
  scale_color_viridis_d() +
  guides(col = guide_legend(title = "Localization"))
ggplot(final.GE) +
  geom_text(aes(comp1, comp2, col = Loc, label = Samples)) +
  ggtitle("Final design", subtitle = "GE") +
  scale_color_viridis_d() +
  guides(col = guide_legend(title = "Localization"))
```

## Conclusions

In general using only two weights leads to select models that account for more 
variance while, even if we can't see in here, the model having only 2 variables 
depends more on the weights regardless which scheme we use.
The weights that report more inner AVE are consistent for each scheme.
The distinction seems to be between adding a slighly relationship between CGH 
and GE or not. [CGE](https://en.wikipedia.org/wiki/Comparative_genomic_hybridization)
allows to find copy number variations, which might affect the gene expression, 
thus is reasonable to include in a model the interaction between copy number 
variation and gene expression, as well as it is unexpected that it might play a .
huge role But we can see that its role is not neglibile 

We can check them:
```{r}
C_model2 <- structure(c(0, 0, 1, 0, 0, 0.1, 1, 0.1, 0), .Dim = c(3L, 3L), .Dimnames = list(
    c("GE", "CGH", "y"), c("GE", "CGH", "y")))
C_model3 <- structure(c(0, 0.1, 1, 0.1, 0, 0.1, 1, 0.1, 0), .Dim = c(3L, 3L), .Dimnames = list(
    c("GE", "CGH", "y"), c("GE", "CGH", "y")))

centroid.2 <- sgcca(A, C_model2, c1 = c(.071,.2, 1), ncomp = ncomp, 
      scheme = "centroid", verbose = FALSE)
centroid.3 <- sgcca(A, C_model3, c1 = c(.071,.2, 1), ncomp = ncomp, 
      scheme = "centroid", verbose = FALSE)
plot(centroid.2$Y[[1]][, 1], centroid.2$Y[[2]][, 1], 
     col = Loc, pch = 16, xlab = "GE", ylab = "CGH", main = "Model 2")
plot(centroid.3$Y[[1]][, 1], centroid.3$Y[[2]][, 1], 
     col = Loc, pch = 16, xlab = "GE", ylab = "CGH", main = "Model 3")
```

Note that there are few differences between them, both of them allow to 
differentiate localization of the tumor by gene expression. Indeed, both of them
select the same genes `r length(intersect(names(comp.GE.3)[comp.GE.3 != 0], names(comp.GE.2)[comp.GE.2 != 0]))`,
but select different number of copy number variations (`r sum(comp.CGH.2 != 0)` and `r sum(comp.CGH.3 != 0)`, for model 2 and model 3), altough `r ength(intersect(names(comp.CGH.3)[comp.CGH.3 != 0], names(comp.CGH.2)[comp.CGH.2 != 0]))` are shared. 

## Session Info {-}

```{r session, echo=FALSE}
devtools::session_info()
```
